sdk/python/python.go

```diff
@@ -221,18 +252,101 @@ func ActivateVirtualEnv(environ []string, virtualEnvDir string) []string {
 }
 
 // InstallDependencies will create a new virtual environment and install dependencies in the root directory.
-func InstallDependencies(ctx context.Context, root, venvDir string, showOutput bool) error {
-	return InstallDependenciesWithWriters(ctx, root, venvDir, showOutput, os.Stdout, os.Stderr)
+func InstallDependencies(ctx context.Context, root, venvDir string, showOutput bool, toolchain Toolchain) error {
+	switch toolchain {
+		case Venv:
+			return InstallVenvDependenciesWithWriters(ctx, root, venvDir, showOutput, os.Stdout, os.Stderr)
+		case Poetry:
+			return InstallPoetryDependenciesWithWriters(ctx, root, showOutput, os.Stdout, os.Stderr)
+		default:
+			return errors.Errorf("unsupported toolchain %q", toolchain)
+	}
+}
+
+func InstallPoetryDependenciesWithWriters(ctx context.Context,
+	root string, showOutput bool, infoWriter, errorWriter io.Writer) error {
+	print := func(message string) {
+		if showOutput {
+			fmt.Fprintf(infoWriter, "%s", message)
+		}
+	}
+	os.Remove(path.Join(root, "requirements.txt"))
+	print("Creating poetry environment and installing dependencies...")
+	cmd, err := PoetryCommand(ctx, "install")
+	if err != nil {
+		return err
+	}
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		if len(output) > 0 {
+			fmt.Fprintf(errorWriter, "%s", string(output))
+		}
+		return errors.Errorf("failed to install dependencies")
+	}
+	fmt.Fprintf(infoWriter, "%s", string(output))
+	envCmd, err := PoetryCommand(ctx, "env", "info", "--path")
+	if err != nil {
+		return err
+	}
+	envPath, err := envCmd.CombinedOutput()
+	if err != nil {
+		return err
+	}
+	pulumiYamlFilepath := filepath.Join(root, "Pulumi.yaml")
+	filedata, err := os.ReadFile(pulumiYamlFilepath)
+	if err != nil {
+		return err
+	}
+	var workspaceDocument yaml.Node
+	err = yaml.Unmarshal(filedata, &workspaceDocument)
+	if err != nil {
+		return err
+	}
+	err = yamlutil.Delete(&workspaceDocument, "runtime")
+	if err != nil {
+		return err
+	}
+	projectName, found, err := yamlutil.Get(&workspaceDocument, "name")
+	if err != nil {
+		return err
+	}
+	if found {
+		err = yamlutil.Insert(&workspaceDocument, "main", strings.TrimSpace(fmt.Sprintf(`src/%s`, string(projectName.Value))))
+	}
+	if err != nil {
+		return err
+	}
```

Comment:
Is it just a golang thing that every function call is followed by error checking?

Reply:
Golang doesn't approve of exceptions or monads (like `Result<T, Err>`) so welcome to 90's style C `if err != nil` 😭 although at least we get proper error types and not just `int` now.

Reply:
That is tremendously unfortunate.

Reply:
The monadic approach not being supported, not the exception approach.